### PR TITLE
fix: self-healing NAT mappings with request deduplication

### LINKED ISSUE #2502: Upnp port is not re-announced
@marten-seemann There is a serious problem with upnp port announcement.

Turn on the application (Libp2p application where there is upnp announcement). Upnp port is announced on the router. It is announced every minute.
Then turn off the internet connection in the router settings or pull the wire (emulate a real internet disconnection). Then turn on the Internet.
Check upnp ports. There are none. And, even after 5 minutes they do not appear.
This presents a serious problem. Because the stability of the internet connection is very bad all over the world (I checked).

### LINKED ISSUE #3224: nat: stuck in error loop after router restart
## Repro

I have a long running node (Kubo 0.34.0-rc1) that had port forwarding set up via `nat` module.

Once the router in my network got restarted due to unrelated DHCP changes, the go-libp2p in Kubo started logging below error in loop (every 2 seconds):

```
2025-03-07T16:25:50.397Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:25:50.421Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:10.443Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:10.466Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:30.496Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:30.516Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:50.533Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
2025-03-07T16:26:50.552Z	WARN	nat	nat/nat.go:244	failed to establish port mapping: goupnp: error performing SOAP HTTP request: Post "http://192.168.50.1:44902/ctl/IPConn": dial tcp 192.168.50.1:44902: connect: connection refused
```

This looks like a gap in our logic, perhaps we should detect when the same `:port/ctl/IPConn` fails multiple times (indicating the router restarted) and trigger re-discovery to self-heal this?

### Version info

- github.com/libp2p/go-libp2p v0.41.0
  - github.com/huin/goupnp v1.3.0

### PR DESCRIPTION
**Pull request by lidel**
This PR makes NAT port mappings self-healing after router restarts.
It also adds bunch of tests and eliminates duplicate mapping requests that occur on node start.

### 1. Automatic NAT recovery after router restart

Router restarts can cause UPnP (SOAP) services to change their listening ports, leading to "connection refused" errors that permanently break port mappings. 
This is the problem described in https://github.com/libp2p/go-libp2p/issues/3224#issuecomment-2866844723 (cc @sukunrt @MarcoPolo)

We had Kubo users impacted this since forever, impacting their ability to provide data:
- https://github.com/ipfs/kubo/issues/8963
- https://github.com/ipfs/kubo/issues/9759

This PR implements automatic NAT rediscovery that detects consecutive connection failures, triggers rediscovery after 3 failures, and restores all existing port mappings on the new NAT instance.

FWIW I've tested it extensively in my own LAN – see "Demo" at the end.

IIUC this
- Fixes #3224
- Fixes #2502

### 2. Port mapping deduplication

This is something I've discovered while testing (1), so also fixed it in this PR.

Multiple libp2p transports sharing the same port (TCP, QUIC, WebTransport, WebRTC-direct) were causing duplicate NAT mapping requests. The fix adds deduplication in `NAT.AddMapping()` that checks if mapping already exists before making NAT API calls.

In case of Kubo IPFS node:
- Before: 10 (5 TCP + 5 UDP) mapping attempts for the same port
- After: 2 (1 TCP + 1 UDP) mapping attempt (visible at 04:14:06 in the demo)

## Demo

The following log shows automatic NAT recovery in action. The router's UPnP service (miniupnpd) restarts and changes its SOAP/HTTP endpoint from port 45251 to a different port, causing "connection refused" errors. After 3 consecutive failures, go-libp2p automatically rediscovers the NAT and restores all mappings:

```console
$ export GOLOG_LOG_LEVEL="error,nat=debug"
$ ipfs daemon
Initializing daemon...
Kubo version: 0.37.0-dev-710953446-dirty
Repo version: 16
System version: amd64/linux
Golang version: go1.25.0
PeerID: 12D3KooWAFYEQLmxbY5xowmqMAFADRbwRDQiSM1cfbmphdHKbhex
Swarm listening on 127.0.0.1:4001 (TCP+UDP)
Swarm listening on 172.17.0.1:4001 (TCP+UDP)
Swarm listening on 172.18.0.1:4001 (TCP+UDP)
Swarm listening on 192.168.1.102:4001 (TCP+UDP)
Swarm listening on [::1]:4001 (TCP+UDP)
Run 'ipfs id' to inspect announced and discovered multiaddrs of this node.
RPC API server listening on /ip4/127.0.0.1/tcp/5001
WebUI: http://127.0.0.1:5001/webui
Gateway server listening on /ip4/127.0.0.1/tcp/8080
Daemon is ready
2025-08-16T04:14:06.558+0200	INFO	nat	nat/nat.go:67	DiscoverGateway address:192.168.1.1
2025-08-16T04:14:06.559+0200	INFO	nat	nat/nat.go:172	Starting maintenance of port mapping: tcp/4001
2025-08-16T04:14:06.559+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:14:06.668+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 16174
2025-08-16T04:14:06.668+0200	INFO	nat	nat/nat.go:172	Starting maintenance of port mapping: udp/4001
2025-08-16T04:14:06.668+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:14:06.954+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 16174
2025-08-16T04:14:26.559+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:14:26.567+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 16174
2025-08-16T04:14:26.568+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:14:26.573+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 16174
2025-08-16T04:14:46.574+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:14:46.580+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 16174
2025-08-16T04:14:46.580+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:14:46.585+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 16174
2025-08-16T04:15:06.586+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:15:06.596+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 16174
2025-08-16T04:15:06.596+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:15:06.601+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 16174
2025-08-16T04:15:26.601+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:15:26.607+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 16174
2025-08-16T04:15:26.607+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:15:26.612+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 16174
2025-08-16T04:15:46.613+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:16:16.616+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=tcp internal_port=4001 external_port=0
2025-08-16T04:16:16.616+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:16:16.616+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=udp internal_port=4001 external_port=0
2025-08-16T04:16:36.617+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:16:36.617+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=tcp internal_port=4001 external_port=0
2025-08-16T04:16:36.617+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:16:36.618+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=udp internal_port=4001 external_port=0
2025-08-16T04:17:16.619+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:17:16.619+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=tcp internal_port=4001 external_port=0
2025-08-16T04:17:16.619+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:17:16.620+0200	WARN	nat	nat/nat.go:326	NAT port mapping failed: protocol=udp internal_port=4001 external_port=0
2025-08-16T04:17:36.621+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:17:36.654+0200	WARN	nat	nat/nat.go:304	NAT port mapping failed: protocol=tcp internal_port=4001 error="goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:45251/ctl/IPConn\": dial tcp 192.168.1.1:45251: connect: connection refused"
2025-08-16T04:17:36.654+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:17:36.678+0200	WARN	nat	nat/nat.go:304	NAT port mapping failed: protocol=udp internal_port=4001 error="goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:45251/ctl/IPConn\": dial tcp 192.168.1.1:45251: connect: connection refused"
2025-08-16T04:17:56.678+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:17:56.694+0200	WARN	nat	nat/nat.go:304	NAT port mapping failed: protocol=tcp internal_port=4001 error="goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:45251/ctl/IPConn\": dial tcp 192.168.1.1:45251: connect: connection refused"
2025-08-16T04:17:56.694+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:17:56.694+0200	INFO	nat	nat/nat.go:332	NAT rediscovery triggered due to repeated connection failures
2025-08-16T04:17:56.715+0200	WARN	nat	nat/nat.go:304	NAT port mapping failed: protocol=udp internal_port=4001 error="goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:45251/ctl/IPConn\": dial tcp 192.168.1.1:45251: connect: connection refused"
2025-08-16T04:18:00.824+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:18:00.930+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 65008
2025-08-16T04:18:00.930+0200	INFO	nat	nat/nat.go:367	NAT mapping restored after rediscovery: tcp 4001 -> 65008
2025-08-16T04:18:00.930+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:18:01.223+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 65008
2025-08-16T04:18:01.223+0200	INFO	nat	nat/nat.go:367	NAT mapping restored after rediscovery: udp 4001 -> 65008
2025-08-16T04:18:01.223+0200	INFO	nat	nat/nat.go:372	NAT rediscovery successful
2025-08-16T04:18:16.716+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:18:16.722+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 65008
2025-08-16T04:18:16.722+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:18:16.726+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 65008
2025-08-16T04:18:36.728+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: udp/4001
2025-08-16T04:18:36.734+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: udp 4001 -> 65008
2025-08-16T04:18:36.734+0200	DEBUG	nat	nat/nat.go:281	Attempting port map: tcp/4001
2025-08-16T04:18:36.742+0200	DEBUG	nat	nat/nat.go:298	NAT port mapping established: tcp 4001 -> 65008
^C
Received interrupt signal, shutting down...
(Hit ctrl-c again to force-shutdown the daemon.)
2025-08-16T04:18:45.085+0200	INFO	nat	nat/nat.go:196	Stopping maintenance of port mapping: tcp/4001
2025-08-16T04:18:45.097+0200	INFO	nat	nat/nat.go:196	Stopping maintenance of port mapping: udp/4001
ipfs daemon  14.25s user 3.83s system 6% cpu 4:43.00 total
```

Key moments in the log:
- **04:14:06**: Initial NAT discovery succeeds, only 2 mapping attempts (deduplication working)
- **04:15:46**: Router restarts, UPnP service begins failing
- **04:17:36**: Connection refused errors indicate UPnP port changed from :45251
- **04:17:56**: After 3rd failure, automatic rediscovery triggered
- **04:18:00**: NAT rediscovered, mappings restored with new external port 65008
- **04:18:45**: Clean shutdown removes all mappings

Without this fix, the connection refused errors would continue indefinitely, requiring manual daemon restart.

### CODE REVIEW

**Comment by lidel** [comment]
@MarcoPolo @sukunrt I've resolved the conflicts (https://github.com/libp2p/go-libp2p/pull/3367/commits/d4ac0deec307f7ad785aca522d75747eea44e4ba)  caused by the grand log refactor that landed in master. 

Would it be possible to review this bugfix and ship in the next go-libp2p? :pray:

**Comment by MarcoPolo** [comment]
Yep! I'll take a look this week :)

File: p2p/net/nat/nat.go [hunk 1]
```diff
@@ -149,11 +186,14 @@ func (nat *NAT) AddMapping(ctx context.Context, protocol string, port int) error
 func (nat *NAT) RemoveMapping(ctx context.Context, protocol string, port int) error {
 	nat.mappingmu.Lock()
 	defer nat.mappingmu.Unlock()
+	nat.natmu.Lock()
+	defer nat.natmu.Unlock()
```
Comment (MarcoPolo) [inline p2p/net/nat/nat.go:190]: What was the reason to add this?

Comment (MarcoPolo) [reply p2p/net/nat/nat.go:190]: Ah nvm, due to calling DeletePortMapping

File: p2p/net/nat/nat.go [hunk 2]
```diff
@@ -74,19 +87,35 @@ func DiscoverNAT(ctx context.Context) (*NAT, error) {
...
-	mappingmu sync.RWMutex // guards mappings
-	closed    bool
-	mappings  map[entry]int
+	// Port mappings that should persist across NAT instance changes
+	mappingmu sync.RWMutex
+	closed    bool          // protected by mappingmu
+	mappings  map[entry]int // protected by mappingmu
+
+	// Track connection failures for auto-rediscovery
+	consecutiveFailures int  // protected by natmu
```
Comment (MarcoPolo) [inline p2p/net/nat/nat.go]: Might as well bring this to the top next to nat and natmu

File: p2p/net/nat/nat_test.go [hunk 3]
```diff
@@ -76,9 +103,305 @@ func TestAddMappingInvalidPort(t *testing.T) {
...
+	require.NoError(t, n.AddMapping(context.Background(), "tcp", 4001))
+	expectPortMappingSuccess(mockNAT, "udp", 4002, 4002)
+	require.NoError(t, n.AddMapping(context.Background(), "udp", 4002))
+
+	// Simulate connection refused error that happens when router's UPnP port changes
+	errConnectionRefused := errors.New("goupnp: error performing SOAP HTTP request: Post \"http://192.168.1.1:1234/ctl/IPConn\": dial tcp 192.168.1.1:1234: connect: connection refused")
+
+	// Set up expectations for the failures that will trigger rediscovery
+	for i := 0; i < 3; i++ {
+		expectPortMappingFailure(mockNAT, "tcp", 10000+i, errConnectionRefused)
```
Comment (MarcoPolo) [inline p2p/net/nat/nat_test.go:199]: nit: It would be nice if this expectation was placed right before the call to n.establishMapping so that it's clear they are tied

Comment (MarcoPolo) [reply p2p/net/nat/nat_test.go:199]: same comment for similar patterns below

**Review by MarcoPolo — APPROVED** [review]
minor nits. Looks good. Thanks for this and for testing it!

**Comment by MarcoPolo** [comment]
I’ll deal with the conflicts and merge tomorrow

**Comment by lthibault** [comment]
Is self-healing NAT a planned feature for Rust?

Happy to begin working on a port if there's interest.

**Comment by mishmosh** [comment]
@jxs might be able to comment on how self-healing NAT fits into rust-libp2p plans.
